src/components/UpsertFolder.vue

```diff
@@ -0,0 +1,125 @@
+<script>
+import { stringify } from 'qs';
+import GirderBreadcrumb from './Breadcrumb.vue';
+import GirderMarkdownEditor from './MarkdownEditor.vue';
+
+const GIRDER_FOLDER_ENDPOINT = 'folder';
+
+export default {
+  components: {
+    GirderBreadcrumb,
+    GirderMarkdownEditor,
+  },
+  props: {
+    location: {
+      type: Object,
+      required: true,
+      validator: val => val.type && val.id,
+    },
+    edit: {
+      type: Boolean,
+      default: false,
+    },
+  },
+  inject: ['girderRest'],
+  data() {
+    return {
+      name: '',
+      description: '',
+      error: '',
+    };
+  },
+  computed: {
+    append() {
+      return this.edit ? [] : [`[ ${this.name || 'New Folder'} ]`];
+    },
+  },
+  mounted() {
+    if (this.edit) {
+      this.loadFolder(this.location.id);
+    }
+  },
+  methods: {
+    async upsert() {
+      this.error = '';
+      try {
+        if (this.edit) {
+          await this.girderRest.put(
+            `${GIRDER_FOLDER_ENDPOINT}/${this.location.id}`,
+            stringify({
+              name: this.name,
+              description: this.description,
+            }),
+          );
+        } else {
+          await this.girderRest.post(
+            GIRDER_FOLDER_ENDPOINT,
+            stringify({
+              parentType: this.location.type,
+              parentId: this.location.id,
+              name: this.name,
+              description: this.description,
+              reuseExisting: false,
+            }),
+          );
+        }
+        this.$emit('done');
+      } catch (err) {
+        this.setError(err);
+      }
+    },
+    async loadFolder(id) {
+      this.error = '';
+      try {
+        const { data } = await this.girderRest.get(`${GIRDER_FOLDER_ENDPOINT}/${id}`);
+        this.name = data.name;
+        this.description = data.description;
+      } catch (err) {
+        this.setError(err);
```

Comment:
I have a question in general here. For this kind of *runtime* exception that is beyond the alibity of this component to handle, should we try to show the error message in the component or yield it upper or global level?

Reply:
I think if loadFolder() throws an error, it could be the server is down, the internet is down, the item is deleted, or wrong id is passed, non of these the UpsertFolder.vue is able to handle. So, I am actually is thinking a global toast as a non graceful way or *$emit('load-error')* to let UpsertFolder's parent have a chance to reload the folder or something. Overall, I am thinking about [practice of try catch](https://softwareengineering.stackexchange.com/questions/270406/is-it-a-good-practice-to-catch-all-exception).

Reply:
I vote for `catch` and `emit`, with some default behavior that presents the error in-component. We could later (or now) add a property to disable the default behavior.

Reply:
When we `emit` we should include the thrown exception as an argument.

Reply:
Added catch/emit behavior for UpsertFolder.